File: src/core/container/docker.rs
use std::fs;
use std::env;
use serde_yaml::{Mapping, Value};
use crate::core::container::custom::{read_qube_yaml, CommandValue, ENVValue};
use serde_json;

pub fn convert_and_run(dockerfile_path: &str) {
    let content = fs::read_to_string(dockerfile_path)
        .unwrap_or_else(|_| panic!("Could not read Dockerfile at {}", dockerfile_path));

    let mut from_line: Option<String> = None;
    let mut ports: Vec<String> = Vec::new();
    let mut cmd_line: Option<String> = None;
    let mut installation_commands: Vec<String> = Vec::new();

    for line in content.lines() {
        let trimmed = line.trim();
        match trimmed.split_whitespace().next() {
            Some("FROM") => {
                from_line = trimmed.split_whitespace().nth(1).map(String::from)
            },
            Some("EXPOSE") => {
                ports.extend(trimmed.split_whitespace().skip(1).map(String::from))
            },
            Some("CMD") => {
                cmd_line = Some(trimmed.trim_start_matches("CMD").trim().to_string());
            },
            Some("ENV") if trimmed.contains("INSTALL_") => {
                if trimmed.contains("INSTALL_NODE=true") {
                    installation_commands.push("curl -fsSL https://deb.nodesource.com/setup_23.x | bash -".to_string());
                    installation_commands.push("apt-get install -y nodejs".to_string());
                }
                if trimmed.contains("INSTALL_RUST=true") {
                    installation_commands.push("curl https://sh.rustup.rs -sSf | sh -s -- -y".to_string());
                }
                if trimmed.contains("INSTALL_PYTHON=true") {
                    installation_commands.push("apt-get install -y python3".to_string());
                }
                if trimmed.contains("INSTALL_GOLANG=true") {
                    installation_commands.push("apt-get install -y golang".to_string());
                }
                if trimmed.contains("INSTALL_JAVA=true") {
                    installation_commands.push("apt-get install -y default-jdk".to_string());
                }
            },
            _ => {}
        }
    }

    let from_value = from_line.expect("Dockerfile does not contain a FROM instruction.");
    let cmd_value = cmd_line.expect("Dockerfile does not contain a CMD instruction.");

    let system = if from_value.to_lowercase().contains("node") {
        "Ubuntu24_NODE".to_string()
    } else {
        from_value.clone()
    };

    let mut cmd_list = installation_commands;
    if cmd_value.starts_with('[') && cmd_value.ends_with(']') {
        match serde_json::from_str::<Vec<String>>(&cmd_value) {
            Ok(mut cmds) => {
                cmd_list.append(&mut cmds);
            },
            Err(_) => {
                cmd_list.push(cmd_value);
            }
        }
    } else {
        cmd_list.push(cmd_value);
    }

    let mut container_map = Mapping::new();
    container_map.insert(Value::String("system".to_string()), Value::String(system));
    if !ports.is_empty() {
        let port_values: Vec<Value> = ports.into_iter().map(Value::String).collect();
        container_map.insert(Value::String("ports".to_string()), Value::Sequence(port_values));
    }
    let cmd_values: Vec<Value> = cmd_list.into_iter().map(Value::String).collect();
    container_map.insert(Value::String("cmd".to_string()), Value::Sequence(cmd_values));
    container_map.insert(Value::String("isolated".to_string()), Value::Bool(false));
    container_map.insert(Value::String("debug".to_string()), Value::Bool(false));

    let mut top_level = Mapping::new();
    top_level.insert(Value::String("container".to_string()), Value::Mapping(container_map));

    let yaml_str = serde_yaml::to_string(&top_level)
        .expect("Failed to convert configuration to YAML");

    fs::write("qube.yml", yaml_str)
        .expect("Failed to write qube.yml");
    println!("Converted Dockerfile to qube.yml successfully.");

    let cwd = env::current_dir()
        .expect("Failed to get current directory")
        .to_string_lossy()
        .to_string();
    let config = read_qube_yaml().expect("Error reading qube.yml");
    let ports_str = config.ports.map(|p| p.join(",")).unwrap_or_else(|| "".to_string());
    let isolated = config.isolated.unwrap_or(false);
    let volumes: Vec<(String, String)> = config.volumes.unwrap_or_default()
        .into_iter()
        .map(|v| (v.host_path, v.container_path))
        .collect();
    let image = config.system.trim().to_string();

    crate::core::container::validate_image(&image).expect("Invalid image provided");

    let container_id = crate::core::container::lifecycle::build_container(None, &cwd, &image);
    let command_str = match config.cmd {
        CommandValue::Single(s) => s,
        CommandValue::List(l) => l.join(" && "),
    };
    let cmd_vec = vec![command_str];
    let env_vars = match config.enviroment {
        Some(ENVValue::Single(s)) => vec![s],
        Some(ENVValue::List(l)) => l,
        None => vec![],
    };

    crate::core::tracking::track_container_named(
        &container_id,
        -1,
        &cwd,
        cmd_vec,
        &image,
        &ports_str,
        isolated,
        &volumes,
        &env_vars,
    );
    eprintln!(
        "\nContainer {} built from Dockerfile conversion. It will be started by the daemon.",
        container_id
    );
}

File: internal/core/container/docker.go
package container

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Voyrox/Qube/internal/core/tracking"
	"github.com/fatih/color"
	"gopkg.in/yaml.v3"
)

type DockerfileConfig struct {
	From        string
	Expose      []string
	Cmd         []string
	Env         map[string]string
	InstallCmds []string
}

type QubeConfig struct {
	System      string         `yaml:"system"`
	Cmd         interface{}    `yaml:"cmd"`
	Ports       []string       `yaml:"ports,omitempty"`
	Environment interface{}    `yaml:"enviroment,omitempty"`
	Isolated    bool           `yaml:"isolated,omitempty"`
	Volumes     []VolumeConfig `yaml:"volumes,omitempty"`
}

type VolumeConfig struct {
	HostPath      string `yaml:"host_path"`
	ContainerPath string `yaml:"container_path"`
}

func ConvertAndRun(dockerfilePath string) error {
	config, err := parseDockerfile(dockerfilePath)
	if err != nil {
		return fmt.Errorf("failed to parse Dockerfile: %w", err)
	}

	qubeConfig := convertToQubeYaml(config)

	dockerfileDir := filepath.Dir(dockerfilePath)
	qubeYamlPath := filepath.Join(dockerfileDir, "qube.yml")

	yamlData, err := yaml.Marshal(qubeConfig)
	if err != nil {
		return fmt.Errorf("failed to marshal YAML: %w", err)
	}

	if err := os.WriteFile(qubeYamlPath, yamlData, 0644); err != nil {
		return fmt.Errorf("failed to write qube.yml: %w", err)
	}

	color.Green("Converted Dockerfile to %s", qubeYamlPath)

	return buildFromQubeYaml(qubeYamlPath, dockerfileDir, qubeConfig)
}

func parseDockerfile(path string) (*DockerfileConfig, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	config := &DockerfileConfig{
		Env:    make(map[string]string),
		Expose: []string{},
		Cmd:    []string{},
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		if strings.HasPrefix(line, "FROM ") {
			config.From = strings.TrimSpace(strings.TrimPrefix(line, "FROM "))
		}

		if strings.HasPrefix(line, "EXPOSE ") {
			ports := strings.Fields(strings.TrimPrefix(line, "EXPOSE "))
			config.Expose = append(config.Expose, ports...)
		}

		if strings.HasPrefix(line, "CMD ") {
			cmdStr := strings.TrimSpace(strings.TrimPrefix(line, "CMD "))

			if strings.HasPrefix(cmdStr, "[") {
				var cmdArray []string
				if err := json.Unmarshal([]byte(cmdStr), &cmdArray); err == nil {
					config.Cmd = cmdArray
				} else {
					config.Cmd = []string{cmdStr}
				}
			} else {
				config.Cmd = []string{cmdStr}
			}
		}

		if strings.HasPrefix(line, "ENV ") {
			envStr := strings.TrimSpace(strings.TrimPrefix(line, "ENV "))
			parts := strings.SplitN(envStr, " ", 2)
			if len(parts) == 2 {
				key := parts[0]
				value := strings.Trim(parts[1], "\"'")
				config.Env[key] = value
			}
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	config.InstallCmds = generateInstallCommands(config.Env)

	return config, nil
}

func generateInstallCommands(env map[string]string) []string {
	var commands []string

	if nodeVersion, ok := env["INSTALL_NODE"]; ok && nodeVersion != "" {
		commands = append(commands,
			"curl -fsSL https://deb.nodesource.com/setup_lts.x | bash -",
			"apt-get install -y nodejs",
		)
		if nodeVersion != "latest" {
			commands = append(commands, fmt.Sprintf("npm install -g n && n %s", nodeVersion))
		}
	}

	if rustVersion, ok := env["INSTALL_RUST"]; ok && rustVersion != "" {
		commands = append(commands,
			"curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y",
			"source $HOME/.cargo/env",
		)
		if rustVersion != "latest" {
			commands = append(commands, fmt.Sprintf("rustup install %s", rustVersion))
		}
	}

	if pythonVersion, ok := env["INSTALL_PYTHON"]; ok && pythonVersion != "" {
		if pythonVersion == "latest" || pythonVersion == "3" {
			commands = append(commands, "apt-get update && apt-get install -y python3 python3-pip")
		} else {
			commands = append(commands, fmt.Sprintf("apt-get update && apt-get install -y python%s python%s-pip", pythonVersion, pythonVersion))
		}
	}

	if goVersion, ok := env["INSTALL_GOLANG"]; ok && goVersion != "" {
		if goVersion == "latest" {
			goVersion = "1.21.0"
		}
		commands = append(commands,
			fmt.Sprintf("wget https://go.dev/dl/go%s.linux-amd64.tar.gz", goVersion),
			fmt.Sprintf("tar -C /usr/local -xzf go%s.linux-amd64.tar.gz", goVersion),
			"export PATH=$PATH:/usr/local/go/bin",
		)
	}

	if javaVersion, ok := env["INSTALL_JAVA"]; ok && javaVersion != "" {
		if javaVersion == "latest" || javaVersion == "11" {
			commands = append(commands, "apt-get update && apt-get install -y openjdk-11-jdk")
		} else {
			commands = append(commands, fmt.Sprintf("apt-get update && apt-get install -y openjdk-%s-jdk", javaVersion))
		}
	}

	return commands
}

func convertToQubeYaml(config *DockerfileConfig) QubeConfig {
	qubeConfig := QubeConfig{
		System:   config.From,
		Isolated: true,
	}

	if len(config.Expose) > 0 {
		qubeConfig.Ports = config.Expose
	}

	if len(config.Cmd) > 0 {
		if len(config.InstallCmds) > 0 {
			allCmds := append(config.InstallCmds, config.Cmd...)
			qubeConfig.Cmd = allCmds
		} else if len(config.Cmd) == 1 {
			qubeConfig.Cmd = config.Cmd[0]
		} else {
			qubeConfig.Cmd = config.Cmd
		}
	}

	if len(config.Env) > 0 {
		var envVars []string
		for key, value := range config.Env {
			envVars = append(envVars, fmt.Sprintf("%s=%s", key, value))
		}
		if len(envVars) == 1 {
			qubeConfig.Environment = envVars[0]
		} else {
			qubeConfig.Environment = envVars
		}
	}

	return qubeConfig
}

func buildFromQubeYaml(yamlPath, workDir string, config QubeConfig) error {
	containerID, err := BuildContainer("", workDir, config.System)
	if err != nil {
		return fmt.Errorf("failed to build container: %w", err)
	}

	var cmdVec []string
	switch v := config.Cmd.(type) {
	case string:
		cmdVec = []string{v}
	case []interface{}:
		for _, cmd := range v {
			if str, ok := cmd.(string); ok {
				cmdVec = append(cmdVec, str)
			}
		}
	case []string:
		cmdVec = v
	}

	var envVars []string
	switch v := config.Environment.(type) {
	case string:
		envVars = []string{v}
	case []interface{}:
		for _, env := range v {
			if str, ok := env.(string); ok {
				envVars = append(envVars, str)
			}
		}
	case []string:
		envVars = v
	}

	var volumes [][2]string
	for _, vol := range config.Volumes {
		volumes = append(volumes, [2]string{vol.HostPath, vol.ContainerPath})
	}

	portsStr := ""
	if len(config.Ports) > 0 {
		portsStr = strings.Join(config.Ports, ",")
	}

	tracking.TrackContainerNamed(
		containerID,
		-1,
		workDir,
		cmdVec,
		config.System,
		portsStr,
		config.Isolated,
		volumes,
		envVars,
	)

	color.Green("Container %s built from Dockerfile conversion. It will be started by the daemon.", containerID)
	return nil
}
